### DISCUSSION TITLE
TemplatedControl doesn't pass the properties I set correctly

### DISCUSSION THREAD

**Discussion by sectlying** [AvaloniaUI/Avalonia#14532] [Q&A] [answered]
I'm new to Avalonia.
Version is 11.0.7.
I'm trying to make a templated control called "ListControl",but I can't get the framework to pass my properties correctly, it just shows the fallback value.

* ListControl.axaml
```axaml
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:controls="using:TsTeC.Templates"
        x:DataType="controls:ListControl"
        xmlns:fluentIcons="clr-namespace:FluentAvalonia.FluentIcons;assembly=FluentAvalonia.FluentIcons"
        xmlns:ns="clr-namespace:TsTeC">
    <Design.PreviewWith>
        <controls:ListControl />
    </Design.PreviewWith>
   
    <Style Selector="controls|ListControl">
        <Setter Property="Template">
            <ControlTemplate>
                <ScrollViewer>
                    <ItemsControl ItemsSource="{Binding Items,RelativeSource={RelativeSource TemplatedParent}}" Name="Control">
                <ItemsControl.Tag>
                    <MultiBinding Converter="{x:Static BoolConverters.Or}">
                        <Binding Path="Client" Converter="{x:Static ObjectConverters.IsNotNull}"/>
                    </MultiBinding>
                </ItemsControl.Tag>
                <ItemsControl.ItemTemplate>
                    <DataTemplate>
                        <Grid ColumnDefinitions="Auto,*,Auto" RowDefinitions="*,Auto" Margin="6" HorizontalAlignment="Stretch">
                            <fluentIcons:FluentIcon Grid.Column="0" Height="24" Width="24" Margin="2" Icon="{Binding Converter={x:Static ns:Converters.IconConverter}}" />
                            <StackPanel Grid.Row="0" Grid.Column="1" Orientation="Vertical">
                                <TextBlock Text="{Binding Name}" Margin="2" />
                                <TextBlock
                                    Text="{Binding Time,Converter={x:Static ns:Converters.TimestampConverter}}"
                                    Margin="2" HorizontalAlignment="Left" />
                                <TextBlock
                                    Text="{Binding Subject,Converter={x:Static ns:Converters.SubjectConverter}}"
                                    Margin="2" HorizontalAlignment="Right" IsVisible="{Binding Overview,RelativeSource={RelativeSource TemplatedParent}}"/>
                            </StackPanel>
                            <SplitButton Grid.Row="0" Grid.Column="2" HorizontalAlignment="Right" Content="Open" Margin="6"
                                         Command="{Binding Open}" IsVisible="{Binding CanOpen}">
                                <SplitButton.Flyout>
                                    <MenuFlyout Placement="Bottom">
                                        <MenuItem Header="Favorate"
                                                  IsVisible="{Binding Favorite,Converter={x:Static BoolConverters.Not}}"
                                                  Command="{Binding SetFavorite}" CommandParameter="True" />
                                        <MenuItem Header="Unfavorite" IsVisible="{Binding Favorite}"
                                                  Command="{Binding SetFavorite}" CommandParameter="False" />
                                        <MenuItem Header="Open Directly" Command="{Binding OpenFile}">
                                            <MenuItem.IsVisible>
                                                <MultiBinding Converter="{x:Static BoolConverters.And}">
                                                    <Binding Path="" RelativeSource="{RelativeSource TemplatedParent}"/>
                                                    <Binding Path="Content"
                                                             Converter="{x:Static StringConverters.IsNotNullOrEmpty}" />
                                                </MultiBinding>
                                            </MenuItem.IsVisible>
                                        </MenuItem>
                                        <MenuItem Header="Open The Folder" Command="{Binding OpenFolder}"
                                                  IsVisible="{Binding IsFileOrImage}" />
                                        <MenuItem Header="ReDownload" Command="{Binding ReDownload}">
                                            <MenuItem.IsVisible>
                                                <MultiBinding Converter="{x:Static BoolConverters.And}">
                                                    <Binding Path="IsFileOrImage" />
                                                     <Binding Path="Tag" RelativeSource="{RelativeSource TemplatedParent}"
                                                              Converter="{x:Static BoolConverters.Not}" /> 
                                                </MultiBinding>
                                            </MenuItem.IsVisible>
                                        </MenuItem>
                                    </MenuFlyout>
                                </SplitButton.Flyout>
                            </SplitButton>
                            <ProgressBar Grid.Row="0" Grid.ColumnSpan="3" Grid.Column="0"/>
                        </Grid>
                    </DataTemplate>
                </ItemsControl.ItemTemplate>
            </ItemsControl>
        </ScrollViewer>
            </ControlTemplate>
        </Setter>
    </Style>
</Styles>

```

* ListControl.axaml.cs
```csharp
using System;
using System.Collections.ObjectModel;
using Avalonia;
using Avalonia.Controls.Primitives;
using TsTeC.Entities;

namespace TsTeC.Templates;

public class ListControl : TemplatedControl
{
    public static readonly StyledProperty<ObservableCollection<Item>> ItemsProperty =
        AvaloniaProperty.Register<ListControl, ObservableCollection<Item>>(nameof(Items),defaultValue:new ObservableCollection<Item>(
            new []{new Item()}));
    public ObservableCollection<Item> Items
    {
        get => GetValue(ItemsProperty);
        set => SetValue(ItemsProperty, value);
    }
    public static readonly StyledProperty<Client?> ClientProperty =
        AvaloniaProperty.Register<ListControl, Client?>(nameof(Client));
    
    public Client? Client
    {
        get => GetValue(ClientProperty);
        set => SetValue(ClientProperty, value);
    }
    public static readonly StyledProperty<bool> OverviewProperty =
        AvaloniaProperty.Register<ListControl, bool>(nameof(Overview),defaultValue:false);
    
    public bool Overview
    {
        get => GetValue(OverviewProperty);
        set => SetValue(OverviewProperty, value);
    }
    public ListControl()
    {
        DataContext = this;
    }
}
```
* MainWindow.axaml(partial)
```axaml
            <templates:ListControl  Grid.Row="1" Items="{Binding SelectedItems}" Client="{Binding Client}" Overview="{Binding SelectedSubject,Converter={x:Static ObjectConverters.IsNull}}"/>
```

* App.axaml(partical)
```axaml
    <Application.Styles>
        <FluentTheme />
        <StyleInclude Source="avares://TsTeC/Templates/ListControl.axaml"/>
    </Application.Styles>
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/68bc1433-bfad-4ef9-97d8-c723bce07da0)
(It's showing the fallback value I've set,but I did provide the right data)

```csharp
Console.WriteLine(SelectedItems.Count);
```
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/d37ddb0a-6055-4cfd-8997-acf73c8c84c0)

What should I do? Thank you very much!

**Comment by gehongyan** [comment]
The `DataContext` of the `ListControl` should not be set to the control itself perhaps.

`DataContext` on `ListControl` is the `MainWindowViewModel` instance. Manual assignment results in a `System.InvalidCastException`.

```
System.InvalidCastException: Unable to cast object of type 'TsTeC.Templates.ListControl' to type 'Sandbox.ViewModels.MainWindowViewModel'.
   at CompiledAvaloniaXaml.XamlIlHelpers.Sandbox.ViewModels.MainWindowViewModel,Sandbox.SelectedItems!Getter(Object)
   at Avalonia.Data.Core.ClrPropertyInfo.Get(Object target)
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.get_Value()
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.SendCurrentValue()
```

Removes the incorrect manual assignment.

```diff
public ListControl()
{
-  DataContext = this;
}
```

Your code shall work.

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/5e14379b-b712-4339-a86e-7cef21ba97b4)

**Reply by sectlying** [reply]
But then I got an exception
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/185fd92d-da7d-4fb0-89c8-86d53def43dc)
```
Unhandled exception. System.InvalidCastException: Unable to cast object of type 'TsTeC.Views.MainWindowViewModel' to type 'Avalonia.AvaloniaObject'.
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.PropertyInfoAccessorFactory.CreateAvaloniaPropertyAccessor(WeakReference`1 target, IPropertyInfo property)
   at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.PropertyInfoAccessorPlugin.Start(WeakReference`1 reference, String propertyName)
   at Avalonia.Data.Core.PropertyAccessorNode.StartListeningCore(WeakReference`1 reference)
   at Avalonia.Data.Core.ExpressionNode.StartListening()
   at Avalonia.Data.Core.ExpressionNode.set_Target(WeakReference`1 value)
   at Avalonia.Data.Core.ExpressionObserver.<StartRoot>b__33_0(Object x)
   at Avalonia.Reactive.AnonymousObserver`1.OnNext(T value)
   at Avalonia.Reactive.Observable.<>c__DisplayClass2_1`2.<Select>b__1(TSource input)
   at Avalonia.Reactive.AnonymousObserver`1.OnNext(T value)
   at Avalonia.Reactive.SingleSubscriberObservableBase`1.PublishNext(T value)
   at Avalonia.Data.BindingBase.UpdateSignal.PropertyChanged(Object sender, AvaloniaPropertyChangedEventArgs e)
   at Avalonia.AvaloniaObject.RaisePropertyChanged[T](AvaloniaProperty`1 property, Optional`1 oldValue, BindingValue`1 newValue, BindingPriority priority, Boolean isEffectiveValue)
   at Avalonia.PropertyStore.EffectiveValue`1.RaiseInheritedValueChanged(AvaloniaObject owner, AvaloniaProperty property, EffectiveValue oldValue, EffectiveValue newValue)
   at Avalonia.PropertyStore.ValueStore.InheritedValueChanged(AvaloniaProperty property, EffectiveValue oldValue, EffectiveValue newValue)
   at Avalonia.PropertyStore.ValueStore.InheritedValueChanged(AvaloniaProperty property, EffectiveValue oldValue, EffectiveValue newValue)
   at Avalonia.PropertyStore.ValueStore.SetInheritanceParent(AvaloniaObject newParent)
   at Avalonia.AvaloniaObject.set_InheritanceParent(AvaloniaObject value)
   at Avalonia.StyledElement.Avalonia.Controls.ISetLogicalParent.SetParent(ILogical parent)
   at Avalonia.Controls.Primitives.TemplatedControl.ApplyTemplate()
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Controls.Grid.MeasureCell(Int32 cell, Boolean forceInfinityV)
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV, Boolean& hasDesiredSizeUChanged)
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV)
   at Avalonia.Controls.Grid.MeasureOverride(Size constraint)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Controls.Grid.MeasureCell(Int32 cell, Boolean forceInfinityV)
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV, Boolean& hasDesiredSizeUChanged)
   at Avalonia.Controls.Grid.MeasureCellsGroup(Int32 cellsHead, Size referenceSize, Boolean ignoreDesiredSizeU, Boolean forceInfinityV)
   at Avalonia.Controls.Grid.MeasureOverride(Size constraint)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding, Thickness borderThickness)
   at Avalonia.Controls.Presenters.ContentPresenter.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutHelper.MeasureChild(Layoutable control, Size availableSize, Thickness padding)
   at Avalonia.Controls.Decorator.MeasureOverride(Size availableSize)
   at Avalonia.Controls.Primitives.VisualLayerManager.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.Layoutable.MeasureOverride(Size availableSize)
   at Avalonia.Controls.Window.MeasureOverride(Size availableSize)
   at Avalonia.Controls.WindowBase.MeasureCore(Size availableSize)
   at Avalonia.Layout.Layoutable.Measure(Size availableSize)
   at Avalonia.Layout.LayoutManager.Measure(Layoutable control)
   at Avalonia.Layout.LayoutManager.ExecuteInitialLayoutPass()
   at Avalonia.Controls.Window.ShowCore(Window owner)
   at Avalonia.Controls.Window.Show()
   at TsTeC.Views.MainWindow.Show() in E:\TsTeC\TsTeC\Views\MainWindow.axaml.cs:line 26
   at TsTeC.App.<>c.<ShowMainWindow>b__0_0() in E:\TsTeC\TsTeC\App.axaml.cs:line 23
   at Avalonia.Threading.Dispatcher.Invoke(Action callback, DispatcherPriority priority, CancellationToken cancellationToken, TimeSpan timeout)
   at Avalonia.Threading.Dispatcher.Invoke(Action callback)
   at TsTeC.App.ShowMainWindow() in E:\TsTeC\TsTeC\App.axaml.cs:line 17
   at TsTeC.App.TrayIcon_OnClicked(Object sender, EventArgs e) in E:\TsTeC\TsTeC\App.axaml.cs:line 48
   at Avalonia.Controls.TrayIcon.<.ctor>b__1_0()
   at Avalonia.Win32.TrayIconImpl.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.TrayIconImpl.ProcWnd(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
   at Avalonia.Win32.Win32Platform.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam)
```
MainWindow.axaml(partical)
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/659fe27a-8cab-4735-ae7a-075238ed7b58)

**Reply by gehongyan** [reply]
I am not able to indicate where your issue takes place from the code provided.

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/2363863d-755f-4488-b712-1d324b052953)

If possible, please provide a minimal reproducible repository.

Additionally, JetBrains Rider supports debugging into the source code. In the debug panel, open `Break Points...` (Visual Studio: Ctrl+Alt+B, Resharper/IntelliJ: Ctrl+Shift+F8), and check `CLR Exception Breakpoints`, to suspend the application when any exception occurs.

![](https://resources.jetbrains.com/help/img/rider/2023.3/debug_new_exception_breakpoint.png)

<details>
<summary>Chinese</summary>
<img src="https://github.com/AvaloniaUI/Avalonia/assets/21241496/df712241-ac44-49df-a956-eede5d2b28ab" />
<img src="https://github.com/AvaloniaUI/Avalonia/assets/21241496/542dbff4-18ab-4a4c-b9d4-8e71a30f743d" />
</details>

Check the value of arguments on `CreateAvaloniaPropertyAccessor` to try finding any useful information about possible incorrect bindings.

**Reply by sectlying** [reply]
Please, can you send me your test code? I'd like to see the correct binding process for the control template.

**Reply by gehongyan** [reply]
I have no idea about your business logic, so all the code I can test with is very simple mock values.

<details>
<summary>Testing code</summary>

```csharp
public class MainWindowViewModel : ViewModelBase
{
    public ObservableCollection<Item> SelectedItems { get; } =
    [
        new Item
        {
            Time = DateTimeOffset.Now,
            Name = "Name1",
            Subject = "Subject1",
            CanOpen = true,
            Favorite = true,
            Content = "Content1",
            IsFileOrImage = true,
        },
        new Item
        {
            Time = DateTimeOffset.Now,
            Name = "Name2",
            Subject = "Subject2",
            CanOpen = false,
            Favorite = false,
            Content = "Content2",
            IsFileOrImage = false,
        },
        new Item
        {
            Time = DateTimeOffset.Now,
            Name = "Name3",
            Subject = "Subject3",
            CanOpen = true,
            Favorite = true,
            Content = "Content3",
            IsFileOrImage = true,
        },
        new Item
        {
            Time = DateTimeOffset.Now,
            Name = "Name4",
            Subject = "Subject4",
            CanOpen = false,
            Favorite = false,
            Content = "Content4",
            IsFileOrImage = false,
        },
    ];
}

public class Item
{
    public DateTimeOffset Time { get; set; } = DateTimeOffset.Now;
    public string Name { get; set; } = "Item";
    public string Subject { get; set; } = "Subject";
    public ICommand Open => ReactiveCommand.Create(() => Console.WriteLine("Open"));
    public bool CanOpen { get; set; } = true;
    public bool Favorite { get; set; } = true;
    public ICommand SetFavorite => ReactiveCommand.Create<string>(arg => Favorite = bool.Parse(arg));
    public ICommand OpenFile => ReactiveCommand.Create(() => Console.WriteLine("OpenFile"));
    public string Content { get; set; } = "Content";
    public ICommand OpenFolder => ReactiveCommand.Create(() => Console.WriteLine("OpenFolder"));
    public bool IsFileOrImage { get; set; } = true;
    public ICommand ReDownload => ReactiveCommand.Create(() => Console.WriteLine("ReDownload"));
}

public static class Converters
{
    public static readonly IconConverter IconConverter = new();
    public static readonly TimestampConverter TimestampConverter = new();
    public static readonly SubjectConverter SubjectConverter = new();
}

public class IconConverter : IValueConverter
{
    /// <inheritdoc />
    public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        FluentIconSymbol.Add20Regular;

    /// <inheritdoc />
    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        throw new NotSupportedException();
}

public class TimestampConverter : IValueConverter
{
    /// <inheritdoc />
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        value is DateTimeOffset dateTime ? dateTime.ToUnixTimeMilliseconds() : null;

    /// <inheritdoc />
    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        throw new NotSupportedException();
}

public class SubjectConverter : IValueConverter
{
    /// <inheritdoc />
    public object? Convert(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        value?.ToString();

    /// <inheritdoc />
    public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture) =>
        throw new NotSupportedException();
}
```
</details>

**Reply by sectlying** [reply]
[TsTeC.zip](https://github.com/AvaloniaUI/Avalonia/files/14210091/TsTeC.zip)
[Lanzou](https://mpcraft.lanzoum.com/iBBdW1nrg6od)
This is a 7z project file (but Github restricts the extension name, so I have to modify it), please help me take a look, thank you!

![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/364e5da5-ce05-4b70-86db-48ab0720cbbe)
If you comment out the 21st line of MainWindow. xaml, it will be able to display the window
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/50966b5e-39f5-4836-9a49-508683894d18)

**Reply by gehongyan** [reply, marked as answer]
The binding error occurs on `Client`.

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/2bcf1a65-1388-463e-814d-54b324fd79fd)

```diff
<ItemsControl.Tag>
    <MultiBinding Converter="{x:Static BoolConverters.Or}">
-        <Binding Path="Client" Converter="{x:Static ObjectConverters.IsNotNull}" />
+        <TemplateBinding Property="Client" Converter="{x:Static ObjectConverters.IsNotNull}" />
    </MultiBinding>
</ItemsControl.Tag>
```

The `Tag` binds to `Client`, which will cause a property matching on the `DataContext` instead of the template. Using `TemplateBinding` could fix the current issue.

If I invoke `RefreshItems()` in the constructor of `MainWindowViewModel`, and add property initializer value for `Item.Time` on `SelectedItems` with `DateTimeOffset.Now.ToUnixTimeSeconds()`, the window shall display as:

![image](https://github.com/AvaloniaUI/Avalonia/assets/21241496/8688e82e-bc10-4292-9322-d6e73154c953)

Remember that:
- If you want to bind to DataContext or other controls, use `Binding`
- If you want to bind the properties on the control itself, use `TemplateBinding`

**Reply by sectlying** [reply]
Thank you,it works now,and I learned a lot from you❤!

But I'm also wondering how this code should be written in the form of TempletedBinding, and how should the properties below be bound if I do so?
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/c08dffe3-7d46-476d-9ad4-c6742a8bb4ab)
![image](https://github.com/AvaloniaUI/Avalonia/assets/77864650/b1318bb8-1caf-4e52-96f2-bc733ddf911b)

**Reply by gehongyan** [reply]
**TL;DR**

```diff
Index: TsTeC/Templates/ListControl.axaml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/TsTeC/Templates/ListControl.axaml b/TsTeC/Templates/ListControl.axaml
--- a/TsTeC/Templates/ListControl.axaml	(revision 5b99737b3ec8e89a5c8932b8139a07e8906310bd)
+++ b/TsTeC/Templates/ListControl.axaml	(date 1707450352178)
@@ -3,7 +3,8 @@
         xmlns:controls="using:TsTeC.Templates"
         x:DataType="controls:ListControl"
         xmlns:fluentIcons="clr-namespace:FluentAvalonia.FluentIcons;assembly=FluentAvalonia.FluentIcons"
-        xmlns:ns="clr-namespace:TsTeC">
+        xmlns:ns="clr-namespace:TsTeC"
+        xmlns:entities="clr-namespace:TsTeC.Entities">
     <Design.PreviewWith>
         <controls:ListControl />
     </Design.PreviewWith>
@@ -12,7 +13,7 @@
         <Setter Property="Template">
             <ControlTemplate>
                 <ScrollViewer>
-                    <ItemsControl ItemsSource="{Binding Items,RelativeSource={RelativeSource TemplatedParent}}"
+                    <ItemsControl ItemsSource="{TemplateBinding Items}"
                                   Name="Control">
                         <ItemsControl.Tag>
                             <MultiBinding Converter="{x:Static BoolConverters.Or}">
@@ -20,7 +21,7 @@
                             </MultiBinding>
                         </ItemsControl.Tag>
                         <ItemsControl.ItemTemplate>
-                            <DataTemplate>
+                            <DataTemplate DataType="entities:Item">
                                 <Grid ColumnDefinitions="Auto,*,Auto" RowDefinitions="*,Auto" Margin="6"
                                       HorizontalAlignment="Stretch">
                                     <fluentIcons:FluentIcon Grid.Column="0" Height="24" Width="24" Margin="2"

```

---

1. The `ItemsControl.ItemsSource` should bind to the Avalonia property defined on `ListControl` itself, hence the usage of `TemplateBinding Items`.
2. In an `ItemsControl`, the `ItemTemplate` dictates how each item in the `ItemsSource` is displayed. Within a `DataTemplate`, typically, the `DataContext` of each element is the respective data item, so `Binding` should be used instead of `TemplateBinding` because they are not binding to the Avalonia properties of the `ListControl`. However, the code analyzer reports errors of not being able to find properties because `DataTemplate` cannot infer the type of data from the non-generic `IEnumerable` type of `ItemsSource`, so you need to manually specify the type information for the `DataContext` of these bindings.

  Regarding how to set the `DataType` in compiled bindings, you can refer to [Compiled Bindings | Avalonia Docs](https://docs.avaloniaui.net/docs/basics/data/data-binding/compiled-bindings).

  The `DataType` property is provided on the `DataTemplate`, and setting its value to the Item class will suffice for all children elements inheriting from it.

**Reply by sectlying** [reply]
Thank you very much. Happy new year! 🎆
